src/mod_cache_users.erl

```diff
@@ -0,0 +1,236 @@
+%% the first anymore because the cleaner has taken action and pushed it behind.
+%% That's fine, worst case this record will live a segment less than expected.
+put_member(Tab, Key, Table) ->
+    EtsSegment = ets:lookup_element(Tab, Table, 2),
+    ets:insert(EtsSegment, {Key}),
+    ok.
+
+del_member(_, _, '$end_of_table') ->
+    ok;
+del_member(Tab, Key, Table) ->
+    case ets:lookup(Tab, Table) of
+        [{_, EtsSegment}] ->
+            ets:delete(EtsSegment, Key),
+            del_member(Tab, Key, ets:next(Tab, Table));
+        [] ->
+            del_member(Tab, Key, ets:next(Tab, Table))
+    end.
+
+del_domain(_, _, '$end_of_table') ->
+    ok;
+del_domain(Tab, Domain, Table) ->
+    case ets:lookup(Tab, Table) of
+        [{_, EtsSegment}] ->
+            ets:match_delete(EtsSegment, {{Domain, '_'}}),
+            del_domain(Tab, Domain, ets:next(Tab, Table));
+        [] ->
+            del_domain(Tab, Domain, ets:next(Tab, Table))
+    end.
+
+-spec start_cache(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start_cache(HostType, Opts) ->
+    Tab = tbl_name(HostType),
+    Spec = #{id => Tab, start => {?MODULE, start_link, [Tab, Opts]},
+             restart => permanent, shutdown => 5000,
+             type => worker, modules => [?MODULE]},
+    {ok, _} = ejabberd_sup:start_child(Spec).
+
+-spec stop_cache(mongooseim:host_type()) -> any().
+stop_cache(HostType) ->
+    ok = ejabberd_sup:stop_child(tbl_name(HostType)).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-record(cache_state, {tab :: atom()}).
+
+purge_tables(#cache_state{tab = Tab}) ->
+    First = ets:first(Tab),
+    Last = ets:last(Tab),
+    EtsSegment = case ets:info(Tab, size) of
+                     1 ->
+                         ets:lookup_element(Tab, First, 2);
+                     _ ->
+                         [{_, E}] = ets:take(Tab, First),
+                         true = ets:insert_new(Tab, {Last + 1, E}),
+                         E
+                 end,
+    ets:delete_all_objects(EtsSegment),
+    ok.
+
+-spec start_link(atom(), gen_mod:module_opts()) -> {ok, pid()}.
+start_link(ParentTab, Opts) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [ParentTab, Opts], []).
+
+init([ParentTab, Opts]) ->
+    TTL = gen_mod:get_opt(ttl, Opts, infinity),
+    N   = gen_mod:get_opt(number_of_segments, Opts, 1),
+    EtsOpts = [ordered_set, named_table, protected, {read_concurrency, true}],
+    ParentTab = ets:new(ParentTab, EtsOpts),
+    lists:foreach(
+      fun(I) ->
+              SegmentOpts = [set, public, {read_concurrency, true}, {write_concurrency, true}],
+              Ref = ets:new(ParentTab, SegmentOpts),
+              ets:insert(ParentTab, {I, Ref})
+      end, lists:seq(1, N)),
+    erlang:send_after(TTL, self(), {purge, TTL}),
+    {ok, #cache_state{tab = ParentTab}}.
```

Comment:
It's a good idea to keep ParentTab name instead of tab/Tab everywhere. Even if it's a longer version.